syscall/devices/hostepoll/hostepoll.c

```diff
@@ -457,26 +465,32 @@ static int _epoll_wait(
         if (retval > maxevents)
             OE_RAISE_ERRNO(OE_EINVAL);
 
+        locked = true;
+        oe_mutex_lock(&epoll->lock);
+
         for (int i = 0; i < retval; i++)
         {
-            struct oe_epoll_event* event = &events[i];
-            const mapping_t* mapping;
+            struct oe_epoll_event* const event = &events[i];
+            const mapping_t* const mapping = _map_find(epoll, event->data.fd);
 
-            oe_spin_lock(&epoll->lock);
+            if (mapping)
+                event->data.u64 = mapping->event.data.u64;
+            else
             {
-                if ((mapping = _map_find(epoll, event->data.fd)))
-                    event->data.u64 = mapping->event.data.u64;
+                // fd has been deleted between the return of epoll_wait and the
+                // acquisition of the lock.
```

Comment:
I think in this case we should still return all the events. When the wait was triggered the event fd was still registered. Let me know your thoughts. Are there any detrimental effects of doing this?

Reply:
Problem is that we cannot assign the real event data (`event->data.u64 = mapping->event.data.u64`) if the fd has been deleted.

Reply:
oh good point. I don't love shifting the whole array left for every deleted event. Instead can we replace the current element with the last element? Does ordering of the events matter to the caller?

Reply:
If order does matter its probably not a huge deal. This is pretty unlikely to happen and very unlikely to happen more than once. But for an application that calls epoll a lot on a lot of handles it could make a difference.

Reply:
AFAIK order does not matter. I agree replacing with last element is better.

Reply:
looks good, thanks